magpie/permissions.py

```diff
@@ -204,6 +260,113 @@ def ace(self, user_or_group):
+            - :func:`magpie.api.management.users.user_utils.combine_user_group_permissions`
+            - :meth:`PermissionSet.__lt__`
+        """
+        if not isinstance(permission1, PermissionSet):
+            permission1 = PermissionSet(permission1)
+        if not isinstance(permission2, PermissionSet):
+            permission2 = PermissionSet(permission2)
+        # both permissions must contain the user or group reference from the original permission tuple to allow compare
+        # they must also have the same permission name to actually resolving the one to preserve
+        if permission1.name != permission2.name or not (permission1.perm_tuple and permission2.perm_tuple):
+            raise ValueError("Invalid resolution attempt between two incomparable permissions.")
+        # when resolving (local inherited resolution), only one user permission on same resource is possible (by design)
+        # hierarchical/effective resolution of resources can differ
+        if context == PermissionType.INHERITED and (
+                (permission1.perm_tuple.resource is not permission2.perm_tuple.resource) or
+                (permission1.type == PermissionType.DIRECT and permission2.type == PermissionType.DIRECT)):
+            raise ValueError("Invalid inherited resolution attempt expected same resources but contain invalid values.")
+
+        # user direct permission always have priority
+        if permission1.type == PermissionType.DIRECT:
+            return permission1
+        if permission2.type == PermissionType.DIRECT:
+            return permission2
+        # when only comparing groups, priority dictates the result
+        priority1 = permission1.group_priority
+        priority2 = permission2.group_priority
+        if priority1 > priority2:
+            return permission1
+        if priority1 < priority2:
+            return permission2
+        # same group priority are resolved according to corresponding permission names/access/scope (__lt__)
+        if permission1 == permission2:
+            # if the two different groups have the exact same resolution value,
+            # indicate that multiple groups resolve into the same access
+            permission1.reason = PERMISSION_REASON_MULTIPLE
+            return permission1  # preserved group in perm-tuple doesn't matter as they are equivalent
+        # otherwise return whichever group permission has higher resolution value
+        return permission2 if permission1 < permission2 else permission1
+
+    @property
+    def group_priority(self):
+        # type: () -> Optional[Union[int, Type[math.inf]]]
+        """
+        Priority accessor in case of group inherited permission resolved by :class:`PermissionTuple`.
+        """
+        if self._tuple is not None and self.type == PermissionType.INHERITED:
+            if self._tuple.group.group_name == get_constant("MAGPIE_ANONYMOUS_GROUP"):
+                return -1
```

Comment:
This special handling answers my question on the database upgrade. But if anonymous and administrator groups were set in the db you would not have to handle them here. Unless those groups are not even in the db?

Reply:
They are in the db and would indeed cause misbehavior. This is why I just didn't write anything at all in the db.
I feel even more like reverting the migration. It will be just a plain property against the instance group-name.